sys/drv/bcm2835_gpio.c

```diff
@@ -71,3 +88,86 @@ void bcm2835_gpio_set_high_detect(resource_t *r, unsigned pin, bool enable) {
     val &= ~mask;
   bus_write_4(r, BCM2835_GPIO_GPHEN(reg), val);
 }
+
+static int bcm2835_gpio_configure_pin(device_t *dev, uint32_t pin,
+                                      uint32_t fsel, uint32_t pull,
+                                      uint32_t intr_detect) {
+  bcm2835_gpio_t *gpio = (bcm2835_gpio_t *)dev->state;
+
+  bcm2835_gpio_function_select(gpio->gpio, pin, (bcm2835_gpio_func_t)fsel);
+  bcm2835_gpio_set_pull(gpio->gpio, pin, (bcm2838_gpio_gppud_t)pull);
+  bcm2835_gpio_set_high_detect(gpio->gpio, pin, intr_detect);
+
+  return 0;
+}
+
+static int bcm2835_gpio_read_fdt_entry(device_t *dev, phandle_t node) {
+  int result = 0;
+  uint32_t *pin_cfgs, *function_cfgs, *pull_cfgs, *intr_detect_cfgs;
+
+  ssize_t pin_cnt = FDT_getencprop_alloc_multi(node, "pins", sizeof(*pin_cfgs),
+                                               (void **)&pin_cfgs);
+  ssize_t function_cnt = FDT_getencprop_alloc_multi(
+    node, "function", sizeof(*function_cfgs), (void **)&function_cfgs);
+  ssize_t pull_cnt = FDT_getencprop_alloc_multi(
+    node, "pull", sizeof(*pull_cfgs), (void **)&pull_cfgs);
+  ssize_t intr_detect_cnt = FDT_getencprop_alloc_multi(
+    node, "intr_detect", sizeof(*intr_detect_cfgs), (void **)&intr_detect_cfgs);
+
+  if (pin_cnt == FDT_PINS_INVAL) {
+    klog("Warning: GPIO FDT entry with no pins property");
+    result = EINVAL;
+    goto bcm2835_gpio_read_fdt_entry_cleanup;
+  }
+
+  for (ssize_t i = 0; i < pin_cnt; i++) {
+    uint32_t pin = pin_cfgs[i];
+    uint32_t fsel =
+      i < function_cnt ? function_cfgs[i] : function_cfgs[function_cnt - 1];
+    uint32_t pull = i < pull_cnt ? pull_cfgs[i] : pull_cfgs[pull_cnt - 1];
+    uint32_t intr_detect = i < intr_detect_cnt
+                             ? intr_detect_cfgs[i]
+                             : intr_detect_cfgs[intr_detect_cnt - 1];
+
+    bcm2835_gpio_configure_pin(dev, pin, fsel, pull, intr_detect);
+  }
+
+bcm2835_gpio_read_fdt_entry_cleanup:
```

Comment:
Since labels always have local scope please rename it to just `cleanup`.

Reply:
Fixed.